extensions-core/kinesis-indexing-service/src/main/java/org/apache/druid/indexing/kinesis/supervisor/KinesisSupervisor.java

```diff
@@ -0,0 +1,507 @@
+      int replicas,
+      String baseSequenceName,
+      ObjectMapper sortingMapper,
+      TreeMap<Integer, Map<String, String>> sequenceOffsets,
+      SeekableStreamIOConfig taskIoConfig,
+      SeekableStreamTuningConfig taskTuningConfig,
+      RowIngestionMetersFactory rowIngestionMetersFactory
+  ) throws JsonProcessingException, NoSuchMethodException, IllegalAccessException, ClassNotFoundException
+  {
+    final String checkpoints = sortingMapper.writerFor(new TypeReference<TreeMap<Integer, Map<String, String>>>()
+    {
+    }).writeValueAsString(sequenceOffsets);
+    final Map<String, Object> context = spec.getContext() == null
+                                        ? ImmutableMap.of(
+        "checkpoints",
+        checkpoints,
+        IS_INCREMENTAL_HANDOFF_SUPPORTED,
+        true
+    ) : ImmutableMap.<String, Object>builder()
+                                            .put("checkpoints", checkpoints)
+                                            .put(IS_INCREMENTAL_HANDOFF_SUPPORTED, true)
+                                            .putAll(spec.getContext())
+                                            .build();
+    List<SeekableStreamIndexTask<String, String>> taskList = new ArrayList<>();
+    for (int i = 0; i < replicas; i++) {
+      String taskId = Joiner.on("_").join(baseSequenceName, getRandomId());
+      taskList.add(new KinesisIndexTask(
+          taskId,
+          new TaskResource(baseSequenceName, 1),
+          spec.getDataSchema(),
+          (KinesisTuningConfig) taskTuningConfig,
+          (KinesisIOConfig) taskIoConfig,
+          context,
+          null,
+          null,
+          rowIngestionMetersFactory
+      ));
+    }
+    return taskList;
+  }
+
+
+  @Override
+  protected RecordSupplier<String, String> setupRecordSupplier()
+      throws IllegalAccessException, NoSuchMethodException, ClassNotFoundException
+  {
+    KinesisSupervisorIOConfig ioConfig = spec.getIoConfig();
+    KinesisTuningConfig taskTuningConfig = spec.getTuningConfig();
+
+    return new KinesisRecordSupplier(
+        KinesisRecordSupplier.getAmazonKinesisClient(
+            ioConfig.getEndpoint(),
+            ioConfig.getAwsAccessKeyId(),
+            ioConfig.getAwsSecretAccessKey(),
+            ioConfig.getAwsAssumedRoleArn(),
+            ioConfig.getAwsExternalId()
+        ),
+        ioConfig.getRecordsPerFetch(),
+        ioConfig.getFetchDelayMillis(),
+        1,
+        ioConfig.isDeaggregate(),
+        taskTuningConfig.getRecordBufferSize(),
+        taskTuningConfig.getRecordBufferOfferTimeout(),
+        taskTuningConfig.getRecordBufferFullWait(),
+        taskTuningConfig.getFetchSequenceNumberTimeout(),
+        taskTuningConfig.getMaxRecordsPerPoll()
+    );
+
+  }
+
+
+  @Override
+  protected void scheduleReporting(ScheduledExecutorService reportingExec)
+  {
+    // Implement this for Kinesis which uses approximate time from latest instead of offset lag
+/*
```

Comment:
If you want to do this in a follow-up PR, please remove all comments, raise a PR, and add a link here. If not, please implement properly.